android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/Router.kt

```diff
@@ -0,0 +1,272 @@
+
+import android.os.Looper
+import androidx.annotation.CallSuper
+import androidx.annotation.MainThread
+import androidx.annotation.VisibleForTesting
+import com.uber.rib.core.Rib.Companion.configuration
+import java.util.Locale
+import java.util.concurrent.CopyOnWriteArrayList
+
+/**
+ * Responsible for handling the addition and removal of children routers.
+ *
+ * @param <I> type of interactor this router routes.
+ */
+abstract class Router<I : Any> protected constructor(
+  component: InteractorBaseComponent<*>?,
+  interactorArg: I,
+  private val ribRefWatcher: RibRefWatcher,
+  private val mainThread: Thread
+) {
+  private val children: MutableList<Router<*>> = CopyOnWriteArrayList()
+  private val interactor: Interactor<*, *> = interactorArg as Interactor<*, *>
+
+  /** @return the Tag. */
+  internal var tag: String? = null
+    private set
+  private var savedInstanceState: Bundle? = null
+  private var isLoaded = false
+
+  protected constructor(interactor: I, component: InteractorBaseComponent<*>) : this(component, interactor, RibRefWatcher.getInstance(), getMainThread())
+  protected constructor(interactor: I) : this(null, interactor, RibRefWatcher.getInstance(), getMainThread())
+
+  @Suppress("UNCHECKED_CAST")
+  protected fun inject(component: InteractorBaseComponent<*>?) {
+    (component as? InteractorBaseComponent<Interactor<*, *>>)?.inject(interactor)
+  }
+
+  protected open fun attachToInteractor() {
+    interactor.setRouterInternal(this)
+  }
+
+  /** @return the interactor owned by the router. */
+  open fun getInteractor(): I {
+    return interactor as I
+  }
+
+  /**
+   * Dispatch back press to the associated interactor. Do not override this.
+   *
+   * @return TRUE if the interactor handles the back press.
+   */
+  open fun handleBackPress(): Boolean {
+    ribRefWatcher.logBreadcrumb("BACKPRESS", null, null)
+    return interactor.handleBackPress()
+  }
+
+  /** Called after the router has been loaded and initialized.  */
+  @Initializer
+  protected open fun didLoad() {
+  }
+
+  /**
+   * Called when a router is being attached. Router subclasses can perform setup here for anything
+   * that is needed again but is cleaned up in willDetach(). Use didLoad() if the setup is only
+   * needed once.
+   */
+  protected fun willAttach() {}
+
+  /**
+   * Called when a router is being a detached, router subclasses should perform any required clean
+   * up here.
+   */
+  protected open fun willDetach() {}
+
+  /**
+   * Attaches a child router to this router. This method will automatically tag the child router by
+   * its class name to namespace its saved instance state [Bundle] object.
+   *
+   *
+   * If you have multiple children of the same class, use [Router.attachChild] to specify a custom tag.
+   *
+   * @param childRouter the [Router] to be attached.
+   */
+  @JvmOverloads
+  @MainThread
+  internal fun attachChildInternal(childRouter: Router<*>, tag: String = childRouter.javaClass.name) {
```

Comment:
Kotlin does not have package-protected equivalent, so I created an internal method that friend modules can use and marked the other method protected